File: dotnet/Dsa/Graph.cs
using System.Collections;

namespace Dsa;

public class Graph
{
    public static void MainGraph()
    {
        /*
            a -> c
            |    |
            v    v
            b    e
            |
            v
            d -> f

            output: a,b,d,f,c,e
        */
        Hashtable graph;
        
        graph = new Hashtable
        {
            {"a", new List<string>{"c", "b"} },
            {"b", new List<string>{"d"} },
            {"c", new List<string>{"e"} },
            {"d", new List<string>{"f"} },
            {"e", new List<string>{} },
            {"f", new List<string>{} },
        };

        // DepthFirstPrint(graph, "a");
        DepthFirstPrintRecursive(graph, "a"); //output: a,c,d,b,d,f

        graph = new Hashtable
        {
            {"f", new List<string>{"g", "i"} },
            {"g", new List<string>{"h"} },
            {"h", new List<string>{} },
            {"i", new List<string>{"g", "k"} },
            {"j", new List<string>{"i"} },
            {"k", new List<string>{} },
        };

        /*
            f -> g -> h
            |  /
            v /
            i <- j
            |
            v
            k

            output: true
        */

        Console.WriteLine("has path? " + HasPathDepthFirst(graph, "f", "k"));
        Console.WriteLine("(recursive) has path? " + HasPathDepthFirstRecursive(graph, "f", "k"));
        Console.WriteLine("has path breadth? " + HasPathBreadthFirst(graph, "f", "k"));


        List<List<string>> list;

        list = new List<List<string>>
        {
            new List<string>{"i", "j"},
            new List<string>{"k", "i"},
            new List<string>{"m", "k"},
            new List<string>{"k", "l"},
            new List<string>{"o", "n"}
        };

        graph = BuildGraph(list);
        Console.WriteLine("undirect path? " + UndirectPath(graph, "j", "m", new HashSet<string>()));
    }

    public static void DepthFirstPrint(Hashtable graph, string source)
    {
        var stack = new Stack<string>();
        stack.Push(source);

        while (stack.Count > 0)
        {
            var current = stack.Pop();
            Console.WriteLine(current);

            var neighbors = (List<string>)graph[current]!;

            foreach (var n in neighbors)
            {
                stack.Push(n);
            }
        }
    }

    public static void DepthFirstPrintRecursive(Hashtable graph, string source)
    {
        Console.WriteLine(source);
        var neighbors = graph[source] as List<string>;

        foreach(var n in neighbors!)
        {
            DepthFirstPrintRecursive(graph, n);
        }
    }

    // n = number of nodes
    // e = number of edges
    // time complexity: O(e) -> we need to travel to every single edge on the graph
    // space complexity: O(n) -> in the worst case scenario, we are going to have every single node on the stack
    public static bool HasPathDepthFirst(Hashtable graph, string source, string dest)
    {
        var stack = new Stack<string>();
        stack.Push(source);

        while (stack.Count > 0)
        {
            var current = stack.Pop();

            if (current == dest)
            {
                return true;
            }

            foreach(var n in (List<string>)graph[current]!)
            {
                stack.Push(n);
            }
        }

        return false;
    }

    // n = number of nodes
    // e = number of edges
    // time complexity: O(e) -> we need to travel to every single edge on the graph
    // space complexity: O(n) -> in the worst case scenario, we are going to have every single node on the call stack
    public static bool HasPathDepthFirstRecursive(Hashtable graph, string source, string dest)
    {
        if (source == dest)
        {
            return true;
        }
        
        foreach(var n in (graph[source] as List<string>)!)
        {
            if (HasPathDepthFirstRecursive(graph, n, dest))
            {
                return true;
            }
        }

        return false;
    }

    // n = number of nodes
    // e = number of edges
    // time complexity: O(e) -> we need to travel to every single edge on the graph
    // space complexity: O(n) -> in the worst case scenario, we are going to have every single node on the queue
    public static bool HasPathBreadthFirst(Hashtable graph, string source, string dest)
    {
        var queue = new Queue<string>();
        queue.Enqueue(source);

        while (queue.Count > 0)
        {
            var current = queue.Dequeue();

            if (current == dest)
            {
                return true;
            }

            foreach(var n in (List<string>)graph[current]!)
            {
                queue.Enqueue(n);
            }
        }

        return false;
    }

    public static Hashtable BuildGraph(List<List<string>> edges)
    {
        var graph = new Hashtable();

        foreach(var e in edges)
        {
            var a = e.First();
            var b = e.Last();

            if (!graph.ContainsKey(a))
            {
                graph[a] = new List<string>();
            }
            if (!graph.ContainsKey(b))
            {
                graph[b] = new List<string>();
            }

            ((List<string>)graph[a]!).Add(b);
            ((List<string>)graph[b]!).Add(a);
        }

        return graph;
    }

    public static bool UndirectPath(Hashtable graph, string source, string dest, HashSet<string> visited)
    {
        if (source == dest)
        {
            return true;
        }
        if (visited.Contains(source))
        {
            return false;
        }

        visited.Add(source);

        foreach(var n in (List<string>)graph[source]!)
        {
            if (UndirectPath(graph, n, dest, visited))
            {
                return true;
            }
        }

        return false;
    }
}

File: go/graph.go
package main

import (
	"fmt"
	"unsafe"
)

/*
1--0--8
...| /
...5

2--3
| /
4

expected output: 2
*/
// n = # of nodes
// e = # of edges
// time complexity: O(n + e), the outer loop iterates over all nodes O(n), and each edge is traversed
// exacly once when processing neighbors
// space complexity: O(n) in worst case when a graph is a line, all nodes will be pushed into the stack
func connectedComponentsCount(graph map[int][]int) {
	visited := map[int]struct{}{}
	fmt.Printf("visited pointer: %p\n", unsafe.Pointer(&visited))
	count := 0

	for k := range graph {
		_, wasVisited := visited[k]

		if wasVisited {
			continue
		}

		dfs(graph, k, visited)
		count++
	}

	fmt.Printf("total count: %d\n", count)
}

func dfs(graph map[int][]int, current int, visited map[int]struct{}) {
	fmt.Printf("visited pointer: %p\n", unsafe.Pointer(&visited))
	stack := []int{}
	stack = append(stack, current)

	for {
		if len(stack) == 0 {
			break
		}
		current = stack[len(stack)-1]
		stack = stack[:len(stack)-1] //pop

		_, wasVisited := visited[current]

		if wasVisited == true {
			continue
		}

		visited[current] = struct{}{}

		for _, n := range graph[current] {
			stack = append(stack, n)
		}
	}
}

// time and space complexity: same as connectedComponentsCount() above
func connectedComponentsCountRecursive(graph map[int][]int) {
	visited := map[int]struct{}{}
	count := 0

	for node := range graph {
		endedGroup := dfsRecursive(graph, node, visited)

		if endedGroup {
			count++
		}
	}

	fmt.Printf("total count: %d\n", count)
}

func dfsRecursive(graph map[int][]int, node int, visited map[int]struct{}) bool {
	if _, wasVisited := visited[node]; wasVisited == true {
		return false
	}

	visited[node] = struct{}{}

	for _, n := range graph[node] {
		dfsRecursive(graph, n, visited)
	}

	return true
}

func largestComponent(graph map[int][]int) {
	visited := make(map[int]struct{})
	largest := -1

	for k := range graph {
		count := dfsCount(graph, k, visited)

		if count > largest {
			largest = count
		}
	}

	fmt.Printf("largest count: %d\n", largest)
}

func dfsCount(graph map[int][]int, node int, visited map[int]struct{}) int {
	if _, wasVisited := visited[node]; wasVisited == true {
		return 0
	}

	visited[node] = struct{}{}
	count := 1

	for _, n := range graph[node] {
		count += dfsCount(graph, n, visited)
	}

	return count
}

// BFS is the best option here (since DFS can eventualy find the target but can potencially run thru all nodes)
func shortestPath(edges [][]string, source string, dest string) int {
	graph := createGraphFromEdges(edges)
	visited := map[string]struct{}{}

	queue := []struct {
		node     string
		distance int
	}{}
	queue = append(queue, struct {
		node     string
		distance int
	}{node: source, distance: 0}) //enqueue

	for {
		if len(queue) == 0 {
			break
		}

		current := queue[0]
		queue = queue[1:] //dequeue

		if current.node == dest {
			return current.distance
		}

		visited[current.node] = struct{}{}

		for _, neighbour := range graph[current.node] {
			if _, wasVisited := visited[neighbour]; wasVisited == false {
				queue = append(queue, struct {
					node     string
					distance int
				}{node: neighbour, distance: current.distance + 1})
			}
		}
	}

	return -1
}

func createGraphFromEdges(edges [][]string) map[string][]string {
	//convert to a map
	graph := make(map[string][]string)

	for _, e := range edges {
		a := e[0]
		b := e[1]

		if _, exists := graph[a]; exists == false {
			graph[a] = []string{}
		}
		if _, exists := graph[b]; exists == false {
			graph[b] = []string{}
		}

		graph[a] = append(graph[a], b)
		graph[b] = append(graph[b], a)
	}

	return graph
}

/*

i 	[         j
		[w,   l,   w,   w,   l,   w]
		[l,   l,   w,   w,   l,   w]
		[w,   l,   w,   w,   w,   w]
		[w,   w,   w,   l,   l,   w]
		[w,   l,   w,   l,   l,   w]
		[w,   w,   w,   w,   w,   w]
	]
*/
// time complexity: O(r*c) since we need to visit all cells in the matrix
// space complexity: O(r*c) in worst case when the matrix is full of land, all cells will be pushed into the stack (all cells needed to be added on the Set)
func islandCount(matrix [][]string) int {
	visited := map[string]struct{}{}
	count := 0

	for row := range matrix {
		for col := range matrix[row] {
			count += dfsIsland(matrix, row, col, visited)
		}
	}

	return count
}

func dfsIsland(matrix [][]string, row int, col int, visited map[string]struct{}) int {
	if matrix[row][col] == "w" { //water
		return 0
	}

	//land
	current := fmt.Sprintf("%d,%d", row, col)
	if _, wasVisited := visited[current]; wasVisited == true {
		return 0
	}

	visited[current] = struct{}{}

	maxRightPosition := len(matrix[row]) - 1
	if col < maxRightPosition {
		//can move right
		dfsIsland(matrix, row, col+1, visited)
	}

	maxDownPosition := len(matrix) - 1
	if row < maxDownPosition {
		//can move down
		dfsIsland(matrix, row+1, col, visited)
	}

	if col > 0 {
		//can move left
		dfsIsland(matrix, row, col-1, visited)
	}

	return 1
}
